dataland-keycloak-adapter/src/main/kotlin/org/dataland/keycloakAdapter/support/keycloak/KeycloakJwtAuthenticationConverter.kt

```diff
@@ -0,0 +1,32 @@
+package org.dataland.keycloakAdapter.support.keycloak
+
+import org.springframework.core.convert.converter.Converter
+import org.springframework.security.authentication.AbstractAuthenticationToken
+import org.springframework.security.core.GrantedAuthority
+import org.springframework.security.oauth2.jwt.Jwt
+import org.springframework.security.oauth2.server.resource.authentication.JwtAuthenticationToken
+import org.springframework.stereotype.Component
+
+/**
+ * A Converter that converts JWTs to AuthenticationTokens
+ */
+@Component
+class KeycloakJwtAuthenticationConverter(
+    val grantedAuthoritiesConverter: KeycloakGrantedAuthoritiesConverter
+) : Converter<Jwt, AbstractAuthenticationToken> {
+
+    /**
+     * The convert method
+     */
+    override fun convert(jwt: Jwt): JwtAuthenticationToken {
+        val authorities: Collection<GrantedAuthority> = grantedAuthoritiesConverter.convert(jwt)
+        val username = getUsernameFrom(jwt)
+        return JwtAuthenticationToken(jwt, authorities, username)
+    }
+
+    private fun getUsernameFrom(jwt: Jwt): String {
+        return if (jwt.hasClaim("preferred_username")) {
+            jwt.getClaimAsString("preferred_username")
+        } else jwt.getSubject()
+    }
+}
```

Comment:
Eventuell wirst du das nach meinen Anpassungen im Api-Key-Manager noch ändern müssen.
Der Username wird dann gar nicht mehr extrahiert, sondern stattdessen die UserId.

Reply:
Ich würde das gerne umdrehen: Zuerst diese hier nach main mergen - und dann main in 440-api-key-manager

Reply:
Einverstanden